### ISSUE TITLE
Rule Change: no-unused-vars: Should return warning / error for function definition not it usage in recursion fn

### ISSUE THREAD

**Issue by dusan-dragon** [eslint/eslint#16125]
### What rule do you want to change?

no-unused-vars

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
const fetchList = async (): Promise<string[]> => {
  const orderList: string[] = []

  const restOfTheList = await fetchList()
  return [...orderList, ...restOfTheList]
}
```

### What does the rule currently do for this code?

It marks issue on line:
`const restOfTheList = await fetchList()`

with Error:
`'fetchList' is assigned a value but never used.eslint[@typescript-eslint/no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars)`

Witch is really confusing because that is actually usage of the fn. Note that no error is provided for first line

### What will the rule do after it's changed?

Rule should return error only for first list (definition of function)

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

I am aware that the reference inside function does not count as it was also mentioned here: https://github.com/eslint/eslint/issues/12572

However for me as developer it does not make sense and it force me to google what is actually going on.

If the error would be shown for definition it would make a little bit more sense. Ideally I would also add note to the error message that fn referencing itself does not count

**Comment by mdjermanovic** [comment]
Hi @dusan-dragon, thanks for the issue!

> Rule should return error only for first list (definition of function)

The core `no-unused-vars` rule already works this way. You can check in [ESLint Playground](https://eslint.org/play#eyJ0ZXh0IjoiY29uc3QgZmV0Y2hMaXN0ID0gYXN5bmMgKCkgPT4ge1xuICBjb25zdCBvcmRlckxpc3QgPSBbXVxuXG4gIGNvbnN0IHJlc3RPZlRoZUxpc3QgPSBhd2FpdCBmZXRjaExpc3QoKVxuICByZXR1cm4gWy4uLm9yZGVyTGlzdCwgLi4ucmVzdE9mVGhlTGlzdF1cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXVudXNlZC12YXJzIjpbImVycm9yIl19LCJlbnYiOnsiZXM2Ijp0cnVlfX19).

> 'fetchList' is assigned a value but never used.eslint[@typescript-eslint/no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars)

This is actually [`@typescript-eslint/no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars/) plugin rule. You can open an issue in this repository:

https://github.com/typescript-eslint/typescript-eslint

I'm closing this issue since it isn't related to ESLint core rules.

**Comment by dusan-dragon** [comment]
Ahh thank you, my bad... confused the tslint / eslint
